File: 0023-merge-k-sorted-lists/0023-merge-k-sorted-lists.cpp
/**
 * Definition for singly-linked list.
 * struct ListNode {
 *     int val;
 *     ListNode *next;
 *     ListNode() : val(0), next(nullptr) {}
 *     ListNode(int x) : val(x), next(nullptr) {}
 *     ListNode(int x, ListNode *next) : val(x), next(next) {}
 * };
 */

class compare{
    public:
    bool operator()(ListNode* a, ListNode* b)
    {
        return (a->val>b->val);
    }
};

class Solution {
public:
    ListNode* mergeKLists(vector<ListNode*>& lists) {
        if(lists.size()==0)
            return NULL;

        priority_queue<ListNode*, vector<ListNode*>, compare> minHeap;

        for(int i=0; i<lists.size(); i++)
        {
            if(lists[i]!=NULL)
                minHeap.push(lists[i]);
        }

        ListNode* head = NULL;
        ListNode* tail = NULL;
        while(!minHeap.empty())
        {
            ListNode* currNode = minHeap.top();
            minHeap.pop();
            if(head == NULL)
            {
                head = currNode;
                tail = currNode;
                if(tail->next!=NULL)
                    minHeap.push(tail->next);
            }
            else
            {
                tail->next = currNode;
                tail = tail->next;
                if(tail->next!=NULL)
                    minHeap.push(tail->next);
            }
        }
        return head;
    }
};

File: 0023-merge-k-sorted-lists/0023-merge-k-sorted-lists.java
/**
 * Definition for singly-linked list.
 * public class ListNode {
 *     int val;
 *     ListNode next;
 *     ListNode() {}
 *     ListNode(int val) { this.val = val; }
 *     ListNode(int val, ListNode next) { this.val = val; this.next = next; }
 * }
 */
class Solution {
    public ListNode mergeKLists(ListNode[] lists) {
        PriorityQueue<ListNode> minHeap = new PriorityQueue<>((a,b)->a.val-b.val);
        int k = lists.length;
        ListNode head = new ListNode();
        ListNode temp = head;

        for(int i=0; i<k; i++) {
            if(lists[i]!=null)
                minHeap.add(lists[i]);
        }

        while(!minHeap.isEmpty()) {
            ListNode top = minHeap.remove();
            temp.next = top;
            temp = temp.next;
            if(top.next!=null)
                minHeap.add(top.next);
        }

        return head.next;
    }
}